In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

### I started by importing necessaries libraries including pandas for data manipulation and scikit-learn for modeling

In [4]:
data= pd.read_csv(r"C:\Users\Dell\Downloads\temperature.csv")
data

station        Date  Present_Tmax  Present_Tmin  LDAPS_RHmin  \
0         1.0  30-06-2013          28.7          21.4    58.255688   
1         2.0  30-06-2013          31.9          21.6    52.263397   
2         3.0  30-06-2013          31.6          23.3    48.690479   
3         4.0  30-06-2013          32.0          23.4    58.239788   
4         5.0  30-06-2013          31.4          21.9    56.174095   
...       ...         ...           ...           ...          ...   
7747     23.0  30-08-2017          23.3          17.1    26.741310   
7748     24.0  30-08-2017          23.3          17.7    24.040634   
7749     25.0  30-08-2017          23.2          17.4    22.933014   
7750      NaN         NaN          20.0          11.3    19.794666   
7751      NaN         NaN          37.6          29.9    98.524734   

      LDAPS_RHmax  LDAPS_Tmax_lapse  LDAPS_Tmin_lapse   LDAPS_WS    LDAPS_LH  \
0       91.116364         28.074101         23.006936   6.818887   69.451805   
1       90.604721         29.850689         24.035009   5.691890   51.937448   
2       83.973587         30.091292         24.565633   6.138224   20.573050   
3       96.483688         29.704629         23.326177   5.650050   65.727144   
4       90.155128         29.113934         23.486480   5.735004  107.965535   
...           ...               ...               ...        ...         ...   
7747    78.869858         26.352081         18.775678   6.148918   72.058294   
7748    77.294975         27.010193         18.733519   6.542819   47.241457   
7749    77.243744         27.939516         18.522965   7.289264    9.090034   
7750    58.936283         17.624954         14.272646   2.882580  -13.603212   
7751   100.000153         38.542255         29.619342  21.857621  213.414006   

      ...  LDAPS_PPT2  LDAPS_PPT3  LDAPS_PPT4      lat      lon       DEM  \
0     ...    0.000000    0.000000    0.000000  37.6046  126.991  212.3350   
1     ...    0.000000    0.000000    0.000000  37.6046  127.032   44.7624   
2     ...    0.000000    0.000000    0.000000  37.5776  127.058   33.3068   
3     ...    0.000000    0.000000    0.000000  37.6450  127.022   45.7160   
4     ...    0.000000    0.000000    0.000000  37.5507  127.135   35.0380   
...   ...         ...         ...         ...      ...      ...       ...   
7747  ...    0.000000    0.000000    0.000000  37.5372  126.891   15.5876   
7748  ...    0.000000    0.000000    0.000000  37.5237  126.909   17.2956   
7749  ...    0.000000    0.000000    0.000000  37.5237  126.970   19.5844   
7750  ...    0.000000    0.000000    0.000000  37.4562  126.826   12.3700   
7751  ...   21.621661   15.841235   16.655469  37.6450  127.135  212.3350   

         Slope  Solar radiation  Next_Tmax  Next_Tmin  
0     2.785000      5992.895996       29.1       21.2  
1     0.514100      5869.312500       30.5       22.5  
2     0.266100      5863.555664       31.1       23.9  
3     2.534800      5856.964844       31.7       24.3  
4     0.505500      5859.552246       31.2       22.5  
...        ...              ...        ...        ...  
7747  0.155400      4443.313965       28.3       18.1  
7748  0.222300      4438.373535       28.6       18.8  
7749  0.271300      4451.345215       27.8       17.4  
7750  0.098475      4329.520508       17.4       11.3  
7751  5.178230      5992.895996       38.9       29.8  

[7752 rows x 25 columns]

In [5]:
column_names = data.columns
for column in column_names:
        print(column + ' - ' + str(data[column].isnull().sum()))

station - 2
Date - 2
Present_Tmax - 70
Present_Tmin - 70
LDAPS_RHmin - 75
LDAPS_RHmax - 75
LDAPS_Tmax_lapse - 75
LDAPS_Tmin_lapse - 75
LDAPS_WS - 75
LDAPS_LH - 75
LDAPS_CC1 - 75
LDAPS_CC2 - 75
LDAPS_CC3 - 75
LDAPS_CC4 - 75
LDAPS_PPT1 - 75
LDAPS_PPT2 - 75
LDAPS_PPT3 - 75
LDAPS_PPT4 - 75
lat - 0
lon - 0
DEM - 0
Slope - 0
Solar radiation - 0
Next_Tmax - 27
Next_Tmin - 27


In [6]:
data = data.drop(columns=['Date'], axis=1)

In [7]:
data.dropna(inplace=True)

In [8]:
data.isnull().sum()

station             0
Present_Tmax        0
Present_Tmin        0
LDAPS_RHmin         0
LDAPS_RHmax         0
LDAPS_Tmax_lapse    0
LDAPS_Tmin_lapse    0
LDAPS_WS            0
LDAPS_LH            0
LDAPS_CC1           0
LDAPS_CC2           0
LDAPS_CC3           0
LDAPS_CC4           0
LDAPS_PPT1          0
LDAPS_PPT2          0
LDAPS_PPT3          0
LDAPS_PPT4          0
lat                 0
lon                 0
DEM                 0
Slope               0
Solar radiation     0
Next_Tmax           0
Next_Tmin           0
dtype: int64

### checked missing data in the dataset by iterating through the column names and printing the count of null values for each column

In [13]:
X = data.drop(['Next_Tmax', 'Next_Tmin'], axis=1)
y_max = data['Next_Tmax']
y_min = data['Next_Tmin']

In [14]:
X_train_max, X_test_max, y_train_max, y_test_max = train_test_split(X, y_max, test_size=0.2, random_state=42)
X_train_min, X_test_min, y_train_min, y_test_min = train_test_split(X, y_min, test_size=0.2, random_state=42)


In [15]:
model_max = LinearRegression()
model_max.fit(X_train_max, y_train_max)

LinearRegression()

In [16]:
model_max.score(X_train_max, y_train_max)

0.7837849749269237

In [18]:
y_pred_max = model_max.predict(X_test_max)
mae_max = mean_absolute_error(y_test_max, y_pred_max)

In [19]:
y_pred_min = model_min.predict(X_test_min)
mae_min = mean_absolute_error(y_test_min, y_pred_min)

In [20]:
print(f'MAE for Next_Tmax: {mae_max}')
print(f'MAE for Next_Tmin: {mae_min}')

MAE for Next_Tmax: 1.1176890126587176
MAE for Next_Tmin: 0.7868062654957808


## The R-squared for the Next_Tmax model on the training set is approximately0.7838. This score measure the goodnes of fit, indicating that the model explains about 78.38%

## The MAE for Next_Tmax is approximately 1.1177, which represents the average absolute difference between the predicted and actual values for Next_Tmax

##  * The Random forest regressor model performs better with a trainning data. The test set score is approximately 0.8826, suggesting that the model generalize reasoonably well to new data.